controller/controller_test.go

```diff
@@ -0,0 +1,79 @@
+package controller
+
+import (
+	"testing"
+	"time"
+)
+
+func TestGenerateRandomToken(t *testing.T) {
+	n := 256
+	key, err := GenerateRandomToken(n)
+	if err != nil || len(key) != n {
+		t.Errorf("The signing key is wrong: %s", key)
+	}
+}
+
+func TestGenerateRandomString(t *testing.T) {
+	n := 5
+	s, err := GenerateRandomString(n)
+	if err != nil || len(s) < n {
+		t.Errorf("The signing key is wrong: %s", s)
+	}
+}
+
+func TestCheckUserCredentials(t *testing.T) {
+	cases := []struct {
+		email, pass string
+		exists      bool
+	}{
+		{"user@gmail.com", "S3cure3Pa$$", true},
+		{"user@gmail.com", "wrongpass", false},
+		{"1000@doesntexist.com", "bloop", false},
+	}
+
+	for _, c := range cases {
+		err := CheckUserCredentials(c.email, c.pass)
+		if (err != nil && c.exists) || (err == nil && !c.exists) {
+			t.Errorf("User credential validation failed (email: %s - pass: %s). Error: %v", c.email, c.pass, err)
+		}
+	}
+}
+
+func TestCreateAndValidateJWT(t *testing.T) {
+	token, err := CreateJWT(2)
+	if err != nil {
+		t.Errorf("JWT creation failed: %s", err)
+	}
+	//test validation of valid token
+	_, err = ValidateJWT(token)
+	if err != nil {
+		t.Errorf("Validation failed when it should have succeeded: %v", err)
+	}
+	//test validation of garbage token
+	_, err = ValidateJWT("please fail")
+	if err == nil {
+		t.Errorf("Validation succeeded when it should have failed")
+	}
+	//test expiration of token
+	time.Sleep(3 * time.Second)
```

Comment:
If you need to manipulate time through tests, consider using something like https://github.com/jonboulle/clockwork to pass a clock to the component you need to test. `time.Sleep` in tests is not a good practice, although you could let it stand here if that's the only occurrence. If you decide to leave things as-is, please add a comment.

Reply:
I will add a TODO comment to come back to this if there is time since this is the only instance. For now leaving as is. Thanks for showing me this though, very useful